src/generate.rs

```diff
@@ -0,0 +1,325 @@
+            visit_dirs(&entry_path, families)?;
+        } else if let Some(extension) = entry_path.extension() {
+            if extension == "pdsc" {
+                log::info!("Found .pdsc file: {}", path.display());
+
+                handle_package::<std::fs::File>(
+                    Package::from_path(&entry.path()).map_err(|e| e.compat())?,
+                    Kind::Directory(path),
+                    families,
+                )
+                .context(format!(
+                    "Failed to process .pdsc file {}.",
+                    entry.path().display()
+                ))?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn visit_file(path: &Path, families: &mut Vec<ChipFamily>) -> Result<()> {
+    log::info!("Trying to open pack file: {}.", path.display());
+    // If we get a file, try to unpack it.
+    let file = fs::File::open(&path)?;
+
+    let mut archive = zip::ZipArchive::new(file)?;
+
+    let mut pdsc_file = find_pdsc_in_archive(&mut archive)
+        .ok_or_else(|| anyhow!("Failed to find .pdsc file in archive {}", path.display()))?;
+
+    let mut pdsc = String::new();
+    pdsc_file.read_to_string(&mut pdsc)?;
+
+    let package = Package::from_string(&pdsc).map_err(|e| {
+        anyhow!(
+            "Failed to parse pdsc file '{}' in CMSIS Pack {}: {}",
+            pdsc_file.sanitized_name().display(),
+            path.display(),
+            e
+        )
+    })?;
+
+    drop(pdsc_file);
+
+    handle_package(package, Kind::Archive(&mut archive), families)
+}
+
+pub(crate) fn visit_arm_files(families: &mut Vec<ChipFamily>) -> Result<()> {
+    let packs = crate::fetch::list_packs()?;
+
+    for (i, pack) in packs.iter().enumerate() {
+        log::info!("Working PACK {}/{} ...", i, packs.len());
+        visit_arm_file(families, &pack);
+    }
+
+    Ok(())
+}
+
+pub(crate) fn visit_arm_file(families: &mut Vec<ChipFamily>, pack: &Pack) {
+    let mut url = pack.PackUrl.clone();
+    if !url.starts_with("http") {
+        url = format!("https://keilpack.azureedge.net/pack/{}", url);
+    }
+
+    log::info!("Downloading {}", url);
+
+    let response = match reqwest::blocking::get(&url) {
+        Ok(response) => response,
+        Err(error) => {
+            log::error!("Failed to download pack '{}': {}", url, error);
+            return;
+        }
+    };
+    let bytes = match response.bytes() {
+        Ok(bytes) => bytes,
+        Err(error) => {
+            log::error!("Failed to get bytes from pack '{}': {}", url, error);
+            return;
+        }
+    };
+
+    log::info!("Trying to open pack file: {}.", url);
+    let zip = std::io::Cursor::new(bytes);
+    let mut archive = match zip::ZipArchive::new(zip) {
+        Ok(archive) => archive,
+        Err(error) => {
+            log::error!("Failed to open pack '{}': {}", url, error);
+            return;
+        }
+    };
+
+    let mut pdsc_file = match find_pdsc_in_archive(&mut archive) {
```

Comment:
Could the remainder of this function be shared with the `visit_file` function?

Reply:
I don't think that would make sense since this function is meant not to fail and the `visit_file` function can fail (you don't want to ever fail the former since that would interrupt the entire loop looping over all packs, when there is let's say a broken URL.

Reply:
The function returns an error, so we could just ignore that.